File: src/modules/organization/services/organization_create_service.py
from django.db.models import Q
from src.modules.address.models import Address
from src.modules.organization.models import Organization
from src.modules.organization.domain.dto.organization_create_dto import (
    OrganizationCreateDTO,
)
from src.shared.exceptions.conflict_exception import ConflictException


class OrganizationCreateService:
    @classmethod
    def execute(self, dto: OrganizationCreateDTO):

        organization = Organization.objects.filter(
            (Q(name=dto.name) | Q(email=dto.email)) & Q(deleted_at=None)
        ).first()

        if organization is not None:
            raise ConflictException("Organização já cadastrada!")

        created_organization = Organization.objects.create(
            name=dto.name,
            trade_name=dto.trade_name,
            email=dto.email,
            status=dto.status,
            address=Address.objects.create(
                state=dto.address.state,
                city=dto.address.city,
                street=dto.address.street,
                neighborhood=dto.address.neighborhood,
                complement=dto.address.complement,
            ),
        )

        return created_organization

File: src/modules/organization/services/organization-create.service.ts
import { prisma } from '../../../infra/database/prisma.service';
import { ConflictException } from '../../../shared/exceptions';
import { OrganizationCreateDto } from '../domain/dto/organization-create.dto';
import { OrganizationEntity } from '../domain/entities/organization.entity';

class OrganizationCreateService {
  async execute(dto: OrganizationCreateDto): Promise<OrganizationEntity> {
    const organization = await prisma.organization.findFirst({
      where: {
        OR: [
          {
            name: dto.name,
          },
          {
            email: dto.email,
          },
        ],
        deletedAt: null,
      },
    });

    if (organization) throw new ConflictException('Organização já cadastrada!');
    const createdOrganization = await prisma.organization.create({
      data: {
        name: dto.name,
        tradeName: dto.tradeName,
        email: dto.email,
        status: dto.status,
        address: {
          create: dto.address,
        },
      },
      include: {
        address: true,
      },
    });

    return new OrganizationEntity(createdOrganization);
  }
}

export const organizationCreateService = new OrganizationCreateService();
